tempodb/encoding/vparquet/block_traceql.go

```diff
@@ -1066,45 +1209,52 @@ func (c *batchCollector) KeepGroup(res *parquetquery.IteratorResult) bool {
 
 	// Copy resource-level attributes to the individual spans now
 	for k, v := range resAttrs {
-		for _, span := range spans {
-			if _, alreadyExists := span.Attributes[k]; !alreadyExists {
-				span.Attributes[k] = v
+		for _, span := range c.buffer {
+			if _, alreadyExists := span.attributes[k]; !alreadyExists {
+				span.attributes[k] = v
 			}
 		}
 	}
 
 	// Remove unmatched attributes
-	for _, span := range spans {
-		for k, v := range span.Attributes {
+	for _, span := range c.buffer {
+		for k, v := range span.attributes {
 			if v.Type == traceql.TypeNil {
-				delete(span.Attributes, k)
+				delete(span.attributes, k)
 			}
 		}
 	}
 
-	sp := &traceql.Spanset{
-		Spans: make([]traceql.Span, 0, len(spans)),
-	}
+	var filteredSpans []traceql.Span
 
 	// Copy over only spans that met minimum criteria
 	if c.requireAtLeastOneMatchOverall {
-		for _, span := range spans {
-			if len(span.Attributes) > 0 {
-				sp.Spans = append(sp.Spans, span)
+		for _, span := range c.buffer {
+			if len(span.attributes) > 0 {
+				filteredSpans = append(filteredSpans, span)
+				continue
 			}
+			putSpan(span)
 		}
 	} else {
-		sp.Spans = spans
+		filteredSpans = make([]traceql.Span, 0, len(c.buffer))
+		for _, span := range c.buffer {
+			filteredSpans = append([]traceql.Span(nil), span)
```

Comment:
Not sure I understand - Can you double-check this loop?  It looks like filteredSpans will only ever contain the last span in the buffer.

Reply:
good catch. surprised tests didn't find that